vllm/v1/structured_output/bitmasking_grammar.py

```diff
@@ -0,0 +1,383 @@
+                pin_memory=pin_memory)
+            self.structured_decode_arange = torch.arange(0,
+                                                         32,
+                                                         device="cpu",
+                                                         pin_memory=pin_memory)
+            self.grammar_bitmask_cpu = torch.zeros(
+                (self.max_num_reqs, cdiv(self.tpu_vocab_size, 32)),
+                dtype=torch.int32,
+                device="cpu",
+                pin_memory=pin_memory)
+
+    def prepare_structured_decoding_input_tpu(
+        self, logits: torch.Tensor, scheduler_output: SchedulerOutput,
+        input_batch: InputBatch
+    ) -> tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
+        meta = cast(BitmaskSOBatchMetaData,
+                    scheduler_output.structured_output_meta)
+        grammar_bitmask = meta.grammar_bitmask
+        assert grammar_bitmask is not None
+        num_reqs, _ = logits.shape
+
+        # Reset pre-allocated tensors
+        self.grammar_bitmask_cpu.zero_()
+        self.require_structured_out_cpu.zero_()
+
+        # We receive the structured output bitmask from the scheduler, but the
+        # indices of the requests in the batch may not match the indices of
+        # the bitmask since the scheduler doesn't know how the tpu runner is
+        # ordering the requests in the batch. We need to match the order of
+        # bitmask with the order of requests
+        struct_out_indices: list[int] = []
+        mask_indices: list[int] = []
+        assert scheduler_output.structured_output_meta is not None
+        for req_id in input_batch.req_ids:
+            mask_index = scheduler_output.structured_output_meta.\
+                structured_output_request_ids.get(req_id)
+            if mask_index is None:
+                continue
+            batch_index = input_batch.req_id_to_index[req_id]
+            struct_out_indices.append(batch_index)
+            mask_indices.append(mask_index)
+        self.grammar_bitmask_cpu[struct_out_indices] = torch.from_numpy(
+            grammar_bitmask[mask_indices])
+        # It's not guaranteed that all requests in this batch require
+        # structured output, so create a bool tensor to represent
+        # the requests that need structured output.
+        struct_out_indices = torch.tensor(struct_out_indices, dtype=torch.long)
+        self.require_structured_out_cpu[struct_out_indices] = True
+        return self.require_structured_out_cpu[:num_reqs].to(logits.device), \
+            self.grammar_bitmask_cpu[:num_reqs].to(logits.device), \
+            self.structured_decode_arange.to(logits.device)
+
+    @torch.compile(backend="openxla", fullgraph=True, dynamic=False)
+    def structured_decode_tpu(self, require_struct_decoding: torch.Tensor,
+                              grammar_bitmask: torch.Tensor,
+                              logits: torch.Tensor, arange: torch.Tensor):
```

Comment:
we can keep the structured_decode forward here but the torch.compile annotation needs to be brought back in the runner.